File: tools/servir_dados/servir_dados.js
#!/usr/bin/env node
/**
 * SisMOM - Servidor HTTP local de dados (Node.js).
 *
 * Equivalente Node do servir_dados.py. Use se você tem Node mas não tem
 * Python — ou se já está usando Electron e quer reusar o runtime.
 *
 * Uso:
 *   node servir_dados.js --dir /caminho/pasta --port 8765
 *
 * CORS aberto, MIME correto pra TIF, listing de diretório no browser.
 */
'use strict';

const http = require('http');
const fs   = require('fs');
const path = require('path');

// ─── Args ────────────────────────────────────────────────────────────
const args = process.argv.slice(2);
let port = 8765, dir = '.', bind = '127.0.0.1';
for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '--port' || a === '-p') { port = parseInt(args[++i], 10); }
    else if (a === '--dir'  || a === '-d') { dir = args[++i]; }
    else if (a === '--bind' || a === '-b') { bind = args[++i]; }
    else if (a === '--help' || a === '-h') {
        console.log('Uso: node servir_dados.js [--dir PATH] [--port N] [--bind ADDR]');
        process.exit(0);
    }
}

const ROOT = path.resolve(dir);
try {
    if (!fs.statSync(ROOT).isDirectory()) throw new Error('não é diretório');
} catch (e) {
    console.error(`ERRO: "${ROOT}" ${e.message || 'não é um diretório'}.`);
    process.exit(1);
}

// ─── MIME types ──────────────────────────────────────────────────────
const MIME = {
    '.png': 'image/png',
    '.gif': 'image/gif',
    '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg',
    '.tif': 'image/tiff', '.tiff': 'image/tiff',
    '.json': 'application/json',
    '.geojson': 'application/json',
    '.html': 'text/html; charset=utf-8',
    '.txt':  'text/plain; charset=utf-8',
    '.csv':  'text/csv; charset=utf-8',
    '.svg':  'image/svg+xml',
    '.webmanifest': 'application/manifest+json',
};

function setCors(res) {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, HEAD, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', '*');
    res.setHeader('Cache-Control', 'no-cache');
}

function renderListing(relUrl, items) {
    const rows = items
        .map(it => `<li><a href="${encodeURIComponent(it.name) + (it.dir ? '/' : '')}">` +
                   `${it.dir ? '📁 ' : '📄 '}${it.name}</a></li>`)
        .join('\n');
    return `<!doctype html><meta charset="utf-8">
<title>${relUrl}</title>
<style>body{font-family:ui-monospace,Menlo,monospace;background:#0b1220;color:#e0e6f0;padding:24px}
a{color:#60a5fa;text-decoration:none}a:hover{text-decoration:underline}
ul{list-style:none;padding:0}li{padding:4px 0}h1{font-size:18px;font-weight:700;color:#aab}
</style><h1>📂 ${relUrl}</h1><ul>${rows}</ul>`;
}

// ─── Servidor ────────────────────────────────────────────────────────
const server = http.createServer((req, res) => {
    setCors(res);
    if (req.method === 'OPTIONS') { res.writeHead(200); res.end(); return; }
    if (req.method !== 'GET' && req.method !== 'HEAD') { res.writeHead(405); res.end('Method not allowed'); return; }

    let urlPath;
    try {
        urlPath = decodeURIComponent((req.url || '/').split('?')[0]);
    } catch (_) { res.writeHead(400); res.end('400'); return; }

    // Anti path-traversal: resolve e checa se ainda começa com ROOT
    const target = path.normalize(path.join(ROOT, urlPath));
    if (!target.startsWith(ROOT)) { res.writeHead(403); res.end('403 Forbidden'); return; }

    fs.stat(target, (err, st) => {
        if (err) { res.writeHead(404); res.end('404 Not Found'); return; }
        if (st.isDirectory()) {
            // redireciona p/ adicionar / se faltou
            if (!urlPath.endsWith('/')) {
                res.writeHead(301, { Location: urlPath + '/' }); res.end(); return;
            }
            fs.readdir(target, { withFileTypes: true }, (e, ents) => {
                if (e) { res.writeHead(500); res.end('500'); return; }
                const items = ents
                    .map(d => ({ name: d.name, dir: d.isDirectory() }))
                    .sort((a, b) => (a.dir === b.dir) ? a.name.localeCompare(b.name) : (a.dir ? -1 : 1));
                if (urlPath !== '/') items.unshift({ name: '..', dir: true });
                const html = renderListing(urlPath, items);
                res.setHeader('Content-Type', MIME['.html']);
                res.writeHead(200);
                res.end(html);
            });
            return;
        }
        if (!st.isFile()) { res.writeHead(404); res.end('404'); return; }
        const ext = path.extname(target).toLowerCase();
        res.setHeader('Content-Type', MIME[ext] || 'application/octet-stream');
        res.setHeader('Content-Length', String(st.size));
        if (req.method === 'HEAD') { res.writeHead(200); res.end(); return; }
        res.writeHead(200);
        fs.createReadStream(target).pipe(res);
    });
});

server.on('error', (e) => {
    if (e.code === 'EADDRINUSE') {
        console.error(`\nERRO: porta ${port} já está em uso. Use --port N para outra.`);
    } else if (e.code === 'EACCES') {
        console.error(`\nERRO: sem permissão para abrir porta ${port}.`);
    } else {
        console.error('\nERRO:', e.message);
    }
    process.exit(1);
});

server.listen(port, bind, () => {
    const bar = '='.repeat(62);
    console.log(bar);
    console.log(' GISELE — Servidor local de dados (Node.js)');
    console.log(bar);
    console.log(` Diretório:  ${ROOT}`);
    console.log(` URL base:   http://localhost:${port}/`);
    console.log(` Interface:  ${bind}  (127.0.0.1 = só esta máquina)`);
    console.log(' CORS:       habilitado');
    console.log('');
    console.log(' Use no template do modelo (Configurar > Editar):');
    console.log(`   http://localhost:${port}/Eta3km/{yyyy}/{mm}/{dd}{hh}/`);
    console.log('');
    console.log(' Ctrl+C para parar.');
    console.log(bar);
});

process.on('SIGINT', () => { console.log('\nServidor parado.'); process.exit(0); });

File: tools/servir_dados/servir_dados.py
#!/usr/bin/env python3
"""
SisMOM - Servidor HTTP local de dados.

Serve uma pasta local em http://localhost:PORTA/ com headers CORS abertos
e MIME types corretos para os formatos do SisMOM (PNG, GIF, TIF, GeoJSON).

Uso:
  python servir_dados.py --dir /caminho/da/pasta --port 8765

Args:
  --dir, -d   pasta a servir (default: diretório atual)
  --port, -p  porta TCP (default: 8765)
  --bind, -b  interface (default: 127.0.0.1, só localhost)

Funciona em Safari, Firefox, Chrome, Edge — qualquer browser. Use no
template do modelo do SisMOM:
  http://localhost:8765/Eta3km/{yyyy}/{mm}/{dd}{hh}/
"""
import argparse
import http.server
import mimetypes
import os
import socketserver
import sys

# Mime types específicos do SisMOM (Python 3 às vezes não conhece .tif)
mimetypes.add_type('image/tiff', '.tif')
mimetypes.add_type('image/tiff', '.tiff')
mimetypes.add_type('application/json', '.geojson')


class CORSHandler(http.server.SimpleHTTPRequestHandler):
    """SimpleHTTPRequestHandler com headers CORS abertos."""

    def end_headers(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, HEAD, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.send_header('Cache-Control', 'no-cache')
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(200)
        self.end_headers()

    # Log mais limpo
    def log_message(self, fmt, *args):
        sys.stdout.write("  %s - %s\n" % (self.address_string(), fmt % args))
        sys.stdout.flush()


class ThreadingHTTPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    """TCPServer multi-thread + reuso de endereço (evita 'Address in use' após Ctrl+C)."""
    allow_reuse_address = True
    daemon_threads = True


def main():
    ap = argparse.ArgumentParser(
        description='GISELE - servidor HTTP local de dados',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    ap.add_argument('--dir', '-d', default='.', help='Diretório raiz (default: atual)')
    ap.add_argument('--port', '-p', type=int, default=8765, help='Porta TCP (default: 8765)')
    ap.add_argument('--bind', '-b', default='127.0.0.1',
                    help='Interface (default: 127.0.0.1 = só localhost; "0.0.0.0" expõe na rede)')
    args = ap.parse_args()

    root = os.path.abspath(args.dir)
    if not os.path.isdir(root):
        print(f'ERRO: "{root}" não é um diretório.', file=sys.stderr)
        sys.exit(1)

    os.chdir(root)

    bar = '=' * 62
    print(bar)
    print(' GISELE — Servidor local de dados')
    print(bar)
    print(f' Diretório:  {root}')
    print(f' URL base:   http://localhost:{args.port}/')
    print(f' Interface:  {args.bind}  (127.0.0.1 = só esta máquina)')
    print(' CORS:       habilitado para qualquer origem')
    print('')
    print(' Use no template do modelo (Configurar > Editar):')
    print(f'   http://localhost:{args.port}/Eta3km/{{yyyy}}/{{mm}}/{{dd}}{{hh}}/')
    print('')
    print(' Ctrl+C para parar.')
    print(bar)

    try:
        with ThreadingHTTPServer((args.bind, args.port), CORSHandler) as srv:
            srv.serve_forever()
    except KeyboardInterrupt:
        print('\nServidor parado.')
    except OSError as e:
        msg = str(e).lower()
        if 'in use' in msg or getattr(e, 'errno', None) in (48, 98, 10048):
            print(f'\nERRO: porta {args.port} já está em uso. Use --port para outra.', file=sys.stderr)
            sys.exit(2)
        raise


if __name__ == '__main__':
    main()
